asserts/internal/grouping.go

```diff
@@ -0,0 +1,176 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package internal
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sort"
+)
+
+// Groupings maintain labels to identify membership to one or more groups.
+// Labels are implemented as subsets of integers from 0
+// up to an excluded maximum, where the integers represent the groups.
+// Assumptions:
+//  - most labels are for one group or very few
+//  - a few labels are sparse with more groups in them
+//  - very few comprise the universe of all groups
+type Groupings struct {
+	n        uint
+	maxGroup uint16
+}
+
+// NewGroupings creates a new Groupings supporting labels for membership
+// to up n groups. n must be a positive multiple of 16 and <=65536.
+func NewGroupings(n int) (*Groupings, error) {
+	if n <= 0 || n > 65536 {
+		return nil, fmt.Errorf("n=%d groups is outside of valid range (0, 65536]", n)
+	}
+	if n%16 != 0 {
+		return nil, fmt.Errorf("n=%d groups is not a multiple of 16", n)
+	}
+	return &Groupings{n: uint(n)}, nil
+}
+
+// WithinRange checks whether group is within the admissible range for
+// labeling otherwise it returns an error.
+func (gr *Groupings) WithinRange(group uint16) error {
+	if uint(group) >= gr.n {
+		return fmt.Errorf("group exceeds admissible maximum: %d >= %d", group, gr.n)
+	}
+	return nil
+}
+
+type Grouping struct {
+	size  uint16
+	elems []uint16
+}
+
+func (g Grouping) Copy() Grouping {
+	elems2 := make([]uint16, len(g.elems), cap(g.elems))
+	copy(elems2[:], g.elems[:])
+	g.elems = elems2
+	return g
+}
+
+func (g *Grouping) search(group uint16) (found bool, j uint16) {
+	j = uint16(sort.Search(int(g.size), func(i int) bool { return g.elems[i] >= group }))
+	if j < g.size && g.elems[j] == group {
+		return true, j
+	}
+	return false, j
```

Comment:
I think that the secondary use `j` is non obvious here. Perhaps adding a comment like this would help:
```
// j serves as an index where the non-found element should be inserted at
// to keep the elems array sorted
```

Reply:
I agree the (false, j) return is not super intuitive.
Also, the function could use a doc comment explaining the semantics.